### PR TITLE
Resolves:  Implement backend that automatically logs user in during local development. #12

### PR DESCRIPTION
**Pull request by rhiannonstanford**

### CODE REVIEW

File: backend/DB.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ export const dbReady = new Promise<void>((resolve, reject) => {
   const intervalId = setInterval(() => {
     console.log("Attempting to connect to db");
     sequelize.authenticate().then(
+      // use await here
```
Comment (internettrans) [inline backend/DB.ts]: Is this comment sort of a to-do to switch to async/await? If so, do you want to just go ahead and implement it?

File: backend/DB.ts [hunk 2]
```diff
@@ -32,10 +33,14 @@ export const dbReady = new Promise<void>((resolve, reject) => {
         });
 
         console.log("Running database migrations");
+        // umzug.down({to: 0}) // undo all migrations
+        // .then(() => {
         return umzug.up().then(() => {
+          // then do the migrations
```
Comment (internettrans) [inline backend/DB.ts]: This comment doesn't make sense to me, since by this time in the code the migrations are already complete.

File: backend/DB.ts [hunk 3]
```diff
@@ -32,10 +33,14 @@ export const dbReady = new Promise<void>((resolve, reject) => {
         });
 
         console.log("Running database migrations");
+        // umzug.down({to: 0}) // undo all migrations
```
Comment (internettrans) [inline backend/DB.ts]: Ideally, I think this would be done via a CLI command rather than having to modify the code. For example, `pnpm exec db-migrate down` works in comunidades-unidas-internal. See https://github.com/sequelize/umzug#cli for how this could be done with umzug.

If you can figure out how to do it with a CLI command, let's add that command to the Readme rather than having this code commented out. If not, that's okay too and we can keep this

File: backend/auth/Auth.ts [hunk 4]
```diff
@@ -6,10 +6,28 @@ This is where we will authenticate users.
...
+
+const LOCAL_DEV_USER_EMAIL = "localdev@email.com";
+
+let passportStrategy = new Strategy(async function (email, password, done) {
+  try {
+    let user = await findUser(email, password);
+    if (user) {
+      return done(null, user);
+    } else {
+      return done(null, false);
```
Comment (internettrans) [inline backend/auth/Auth.ts:24]: Hmm I would have expected returning an error here - why are we calling the errback with `null` as the error when we weren't able to login?

```suggestion
      return done(Error("No such user"));
```

File: backend/auth/Auth.ts [hunk 5]
```diff
@@ -21,33 +39,36 @@ router.use(
...
 );
 
 router.get("/login", renderWebApp);
 
 router.use("/", async (req, res, next) => {
   if (req.session && req.session.passport && req.session.passport.user.id) {
     return next();
+  } else if (process.env.IS_RUNNING_LOCALLY) {
+    const localUser = await findOrCreateLocalUser(LOCAL_DEV_USER_EMAIL);
+    loginPromise(localUser, req)
```
Comment (internettrans) [inline backend/auth/Auth.ts]: Let's use await here, since we're already in an async function. Also, it might be best not to send the actual error, including stacktrace, to the browser in the response due to security risks. Showing a random person a stacktrace could give them hints about how to hack the system. In general, I usually only send the full stacktrace to the browser in local development, not deployed (see https://github.com/JustUtahCoders/comunidades-unidas-internal/blob/6ced64d696290d2da4875f7a95e927be6d28a144/backend/server.js#L68-L72)

```suggestion
    try {
      await loginPromise(localUser, req)
      next()
    } catch (err) {
      console.error(err);
      res.status(500).send("Server error during login")
    }
```

File: backend/auth/Auth.ts [hunk 6]
```diff
@@ -21,33 +39,36 @@ router.use(
...
     res.status(302).redirect("/login");
   }
-});
-
-passport.use(
-  new Strategy(function (req, done) {
-    return done(null, { id: "hello" });
-  })
-);
+}); // router.use("/")
```
Comment (internettrans) [inline backend/auth/Auth.ts]: This seems to be left over from something that's no longer relevant.

```suggestion
```

File: backend/auth/Auth.ts [hunk 7]
```diff
@@ -60,3 +81,15 @@ passport.deserializeUser<User>(function (user, done) {
 interface User {
   id: string;
 }
+
+// Created manual promise because util.promisify did not work with req.login
+const loginPromise = (user, request) => {
```
Comment (internettrans) [inline backend/auth/Auth.ts]: What code were you trying with util.promisify? Here's what I would have tried

```js
const login = util.promisify(request.login)
await login(user)
```

File: backend/users/Users.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  });
+
+  let localUser = users.length > 0 ? users[0] : null;
+
+  if (!localUser) {
+    localUser = await sequelize.models.User.create({
+      firstName: "localDevFirstName",
+      lastName: "localDevLastName",
+      email: email,
+      password: "localDevPassword",
```
Comment (internettrans) [inline backend/users/Users.ts:17]: This should be fine since it only works in development (usually putting passwords in code is not good)

File: backend/users/Users.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      googleAuthToken: "localDevGoogleAuthToken",
```
Comment (internettrans) [inline backend/users/Users.ts]: I think I'd prefer `null` for the googleAuthToken, since otherwise our code will think that the local user has a google account but then will fail when it calls the google apis with `localDevGoogleAuthToken` rather than a real auth token.

File: migrations/20210720144838-create-user.cjs [hunk 10]
```diff
@@ -0,0 +1,40 @@
+"use strict";
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    console.log("-------------in the create user migration");
```
Comment (internettrans) [inline migrations/20210720144838-create-user.cjs]: This looks like a debugging statement that is no longer needed

```suggestion
```

File: package.json [hunk 11]
```diff
@@ -50,7 +50,7 @@
     "url-loader": "^4.1.1",
     "webpack": "^5.37.1",
     "webpack-cli": "^4.7.2",
-    "webpack-dev-server": "^3.11.2"
+    "webpack-dev-server": "^4.0.0-beta.3"
```
Comment (internettrans) [inline package.json]: Why this change? If we upgrade to v4, I'd prefer going straight to 4.0.0-rc.0 since 4.0.0-beta.3 has some breaking changes when compared with the newer rc.0

File: backend/DB.ts (hunk 3 above)
Comment (rhiannonstanford) [reply backend/DB.ts]: All the DB.ts code changes were done pairing with Leah to debug the migrations and seeds.  I was having a problem generating and testing a new user and we discovered that we needed to clear the SequelizeMeta table.  I'm happy to figure out the umzug CLI command, but I think it will take some work and maybe should be on a separate ticket?  

The previous comments were done with Leah to help my understanding and I don't think we need them.

File: backend/auth/Auth.ts (hunk 4 above)
Comment (rhiannonstanford) [reply backend/auth/Auth.ts:24]: I might be confused.  I thought, in Passport, failing to authenticate returns false instead of an error.   ?

File: backend/auth/Auth.ts (hunk 5 above)
Comment (rhiannonstanford) [reply backend/auth/Auth.ts]: I like using the await.  But I thought we were in a local development context, no?

File: backend/auth/Auth.ts (hunk 7 above)
Comment (rhiannonstanford) [reply backend/auth/Auth.ts]: That is exactly what I tried.  Got the error: 

```Error logging in:  Error: passport.initialize() middleware not in use```

File: package.json (hunk 11 above)
Comment (rhiannonstanford) [reply package.json]: We used vers 4.0.0-beta.3 as part of debugging to get my docker logging / watcher to work.  Final fix was closing the number of terminals I had open and restarting the computer.  Should I go back to 3.11.2 or 4.0.0-rc.0?

File: backend/DB.ts (hunk 3 above)
Comment (internettrans) [reply backend/DB.ts]: Okay we can do it in a future PR

File: backend/auth/Auth.ts (hunk 4 above)
Comment (internettrans) [reply backend/auth/Auth.ts:24]: You are right - I was mistaken and we should call it with false when authentication fails.

File: backend/auth/Auth.ts (hunk 5 above)
Comment (internettrans) [reply backend/auth/Auth.ts]: Well right now we're developing the software but this code will be deployed to an environment once we start giving away / selling Flax.

File: backend/auth/Auth.ts (hunk 7 above)
Comment (internettrans) [reply backend/auth/Auth.ts]: Hmm that error is called with `req._passport` is not defined. I think I know why - it's related to the default javascript (`this`) of `request.login()` vs `login()`. In javascript, the default value for `this` changes depending on what is to the left of the period of code calling the function. So in `request.login()` the `this` is `request`, but in `login()` there is no period or anything to the left of it, so it will use the global variable as the default `this`.

To fix, bind the promisified version to the request object, which forces the `this` to always be the `request` object:

```js
const login = util.promisify(request.login).bind(request);
await login(user)
```

File: package.json (hunk 11 above)
Comment (internettrans) [reply package.json]: Let's go to 4.0.0-rc.0

File: backend/auth/Auth.ts (hunk 5 above)
Comment (rhiannonstanford) [reply backend/auth/Auth.ts]: Ahh, I see.  That makes sense to me now.  Thank you.

File: backend/auth/Auth.ts (hunk 7 above)
Comment (rhiannonstanford) [reply backend/auth/Auth.ts]: This is great! I will try this fix.  Thank you.

**Review by internettrans — APPROVED** [review]
Looks good, nice work!
